DM-46970: Handle None values in table mode

timj opened the pull request:
If a translation fails we still need to be able to display the table contents.

timj committed:
Handle None values in table mode

If a translation fails we still need to be able to display
the table contents.

timj committed:
Enable test report upload to codecov

timj committed:
Test that table still works with bad translation

timj committed:
Add python 3.13 to build matrix

andy-slac commented on python/astro_metadata_translator/bin/translate.py:
```diff
@@ -144,12 +144,19 @@ def read_file(
+        def _filter_none(info: ObservationInfo, attr: str) -> Any:
+            val = getattr(info, attr)
+            if val is None:
+                return "-"
```
Will that work if the expected column type is numeric and tries to format it with `%f` specification?

timj replied:
Oops. You are right. I need to do the formatting inside this. All my tests were with string types.

timj replied:
I am thinking I should be using a masked `astropy.table` here...

timj replied:
@andy-slac can you take another look. To use QTable I had to reorganize how files are encountered. It's a bit of a hack but the output is nicer:

```
       ObsId        ImgType               Object                           Filter              Detector  ExpTime 
                                                                                                            s    
------------------- ------- --------------------------------- -------------------------------- -------- ---------
            1038843      --                           w2.+2+2                         i.MP9702    ccd02       nan
ct4m20121211t220632    zero                    postflats-BIAS              solid plate 0.0 0.0       S1       0.0
ct4m20150220t004721 science                       Blind15A_03                                g      S29      86.0
ct4m20190402t050618 science                                       VR DECam c0007 6300.0 2600.0      S29     120.0
ct4m20130213t090140 science                 Trifid and Lagoon g DECam SDSS c0001 4720.0 1520.0       S1      30.0
ct4m20130901t060255 science DES supernova hex SN-S1 tiling 22 z DECam SDSS c0004 9260.0 1520.0       S1     200.0
       HSCA04090000 science                          SSP-Wide                            HSC-R     1_07     150.0
       HSCA90402400 science                         STRIPE82L                            HSC-I     1_12      30.0
             849375 science                                D3                                r    ccd99   300.202
            1038843 science                           w2.+2+2                         i.MP9702    ccd02   615.037
       6377 4 g 407 science                              82 S                                g       g4 53.907456
       SUPE00535770 science               Ecliptic Deep Field                           W-S-R+ nausicaa     200.0
```

andy-slac replied:
Right-aligned strings in formatted tables hurt my eyes 🙂 Also in `butler` output.

timj replied:
Here you go:

```
       ObsId        ImgType               Object                           Filter              Detector  ExpTime 
                                                                                                            s    
------------------- ------- --------------------------------- -------------------------------- -------- ---------
1038843             --      w2.+2+2                           i.MP9702                            ccd02       ———
ct4m20121211t220632 zero    postflats-BIAS                    solid plate 0.0 0.0                    S1       0.0
ct4m20150220t004721 science Blind15A_03                       g                                     S29      86.0
ct4m20190402t050618 science                                   VR DECam c0007 6300.0 2600.0          S29     120.0
ct4m20130213t090140 science Trifid and Lagoon                 g DECam SDSS c0001 4720.0 1520.0       S1      30.0
ct4m20130901t060255 science DES supernova hex SN-S1 tiling 22 z DECam SDSS c0004 9260.0 1520.0       S1     200.0
HSCA04090000        science SSP-Wide                          HSC-R                                1_07     150.0
HSCA90402400        science STRIPE82L                         HSC-I                                1_12      30.0
849375              science D3                                r                                   ccd99   300.202
1038843             science w2.+2+2                           i.MP9702                            ccd02   615.037
6377 4 g 407        science 82 S                              g                                      g4 53.907456
SUPE00535770        science Ecliptic Deep Field               W-S-R+                           nausicaa     200.0
```

andy-slac reviewed (approved):
Looks good, one minor question.

timj committed:
Use QTable for table output

This requires that column data be accumulated during file reads
rather than being output as each file is encountered.

andy-slac commented on python/astro_metadata_translator/bin/translate.py:
```diff
@@ -234,12 +227,33 @@ def translate_or_dump_headers(
+            # QTable does not support a masked column of quantity.
+            data = output_columns[c["label"]]
+            mask = [v is None for v in data]
+            if mask:
```
Hmm, did you mean `any(mask)`?

timj replied:
Yes. Sorry.

andy-slac commented on python/astro_metadata_translator/bin/translate.py:
```diff
@@ -68,6 +73,7 @@ def read_file(
     outstream: IO | None = None,
     output_mode: str = "verbose",
     write_heading: bool = False,
+    output_columns: dict | None = None,
```
Can you annotate it as `defaultdict[str, list]` instead of `dict`?

andy-slac reviewed (approved):
Looks good, a couple of questions. One test is failing probably because table header is centered?

timj committed:
Move QTable creation to separate function

timj committed:
Try to dump table output regularly

Now that the output table accumulates rather than having the
row printed instantly, batch up table output and dump it
every few hundred entries. Without this it's possible people
could be waiting a long time with no output appearing and
eating up lots of memory.
